File: mobile/lib/features/hangly_scene/physics/vec2.dart
import 'dart:math' as math;

/// 2D vector with arithmetic and geometric operations for the Hangly Verlet solver.
class Vec2 {
  final double x;
  final double y;

  const Vec2(this.x, this.y);

  static const Vec2 zero = Vec2(0, 0);

  Vec2 operator +(Vec2 other) => Vec2(x + other.x, y + other.y);
  Vec2 operator -(Vec2 other) => Vec2(x - other.x, y - other.y);
  Vec2 operator *(double scale) => Vec2(x * scale, y * scale);
  Vec2 operator /(double divisor) => Vec2(x / divisor, y / divisor);
  Vec2 operator -() => Vec2(-x, -y);

  double get magnitude => math.sqrt(x * x + y * y);
  double get magnitudeSquared => x * x + y * y;

  Vec2 get normalized {
    final len = magnitude;
    return len > 1e-12 ? this / len : Vec2.zero;
  }

  double distanceTo(Vec2 other) => (other - this).magnitude;

  double dot(Vec2 other) => x * other.x + y * other.y;

  /// Rotates this vector counter-clockwise by [radians].
  Vec2 rotated(double radians) {
    final c = math.cos(radians);
    final s = math.sin(radians);
    return Vec2(x * c - y * s, x * s + y * c);
  }

  /// Clamps the vector's length to [maximum] while keeping direction.
  Vec2 limited(double maximum) {
    final len = magnitude;
    if (len <= maximum || len <= 1e-12) {
      return this;
    }
    return this * (maximum / len);
  }

  static Vec2 lerp(Vec2 a, Vec2 b, double t) {
    return Vec2(a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Vec2 &&
          (x - other.x).abs() < 1e-12 &&
          (y - other.y).abs() < 1e-12;

  @override
  int get hashCode => Object.hash(x, y);

  @override
  String toString() => '(${x.toStringAsFixed(4)}, ${y.toStringAsFixed(4)})';
}

File: src/desktop/Hangly.Core/Geometry/Vec2.cs
//
//  Vec2.cs
//  Hangly
//
//  Minimal 2D vector arithmetic for the rope solver.
//

namespace Hangly.Core.Geometry;

/// <summary>
/// The vector type the simulation is written in, standing in for the original's
/// <c>CGPoint</c>.
/// </summary>
/// <remarks>
/// Verlet integration constantly mixes positions and displacements, so both live in
/// one type and no line of the solver pays for a conversion. These operators are the
/// whole of the maths the rope needs.
///
/// A <c>readonly struct</c> rather than a class: the solver allocates nothing per
/// step, which is what keeps a 240 Hz fixed timestep off the garbage collector.
/// </remarks>
public readonly struct Vec2 : IEquatable<Vec2>
{
    public double X { get; }

    public double Y { get; }

    public Vec2(double x, double y)
    {
        X = x;
        Y = y;
    }

    public static Vec2 Zero => new(0, 0);

    public static Vec2 operator +(Vec2 a, Vec2 b) => new(a.X + b.X, a.Y + b.Y);

    public static Vec2 operator -(Vec2 a, Vec2 b) => new(a.X - b.X, a.Y - b.Y);

    public static Vec2 operator *(Vec2 v, double scale) => new(v.X * scale, v.Y * scale);

    public static Vec2 operator /(Vec2 v, double divisor) => new(v.X / divisor, v.Y / divisor);

    /// <summary>Euclidean length.</summary>
    public double Magnitude => Math.Sqrt((X * X) + (Y * Y));

    /// <summary>Length without the square root, for comparisons.</summary>
    public double MagnitudeSquared => (X * X) + (Y * Y);

    /// <summary>Unit vector, or zero for a zero-length vector.</summary>
    public Vec2 Normalized
    {
        get
        {
            double length = Magnitude;
            return length > Precision.UlpOfOne ? this / length : Zero;
        }
    }

    public double DistanceTo(Vec2 other) => (other - this).Magnitude;

    /// <summary>The vector rotated counter-clockwise by <paramref name="radians"/>.</summary>
    public Vec2 Rotated(double radians)
    {
        double cosine = Math.Cos(radians);
        double sine = Math.Sin(radians);
        return new Vec2((X * cosine) - (Y * sine), (X * sine) + (Y * cosine));
    }

    /// <summary>Caps the vector's length at <paramref name="maximum"/>, keeping its direction.</summary>
    public Vec2 Limited(double maximum)
    {
        double length = Magnitude;
        if (length <= maximum || length <= Precision.UlpOfOne)
        {
            return this;
        }

        return this * (maximum / length);
    }

    public bool Equals(Vec2 other) => X.Equals(other.X) && Y.Equals(other.Y);

    public override bool Equals(object? obj) => obj is Vec2 other && Equals(other);

    public override int GetHashCode() => HashCode.Combine(X, Y);

    public static bool operator ==(Vec2 a, Vec2 b) => a.Equals(b);

    public static bool operator !=(Vec2 a, Vec2 b) => !a.Equals(b);

    public override string ToString() => $"({X:0.####}, {Y:0.####})";
}
